README.md
This is a version of sdcc with an additional backend to the Caltech10 CPU. However, it can only compile the simplest C programs.

ChangeLog
2019-05-02 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* device/include/string.h,
	  device/lib/incl.mk,
	  device/lib/memccpy.c,
	  support/regression/tests/memory.c:
	  Implement C2X memccpy().

2019-04-27 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* src/pdk/ralloc.c,
	  src/pdk/gen.c:
	  Improve register packing.

2019-04-27 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* src/pdk/gen.c,
	  src/pdk/peeph.def:
	  Improve workaround for missing addc/subc a, k; improve shifts by 7, 15, etc; fix handling of non-dead a in wide right shifty by literal.

2019-04-24 Philipp Klaus Krause <pkk AT spth.de>

	* doc/sdccman.lyx:
	  Fix note on endianness.

2019-04-21 Philipp Klaus Krause <pkk AT spth.de>

	* src/pdk/gen.c:
	  Fix bug #2904.
	* src/z80/gen.c:
	  Implement RFE #623.

2019-04-18 Philipp Klaus Krause <pkk AT spth.de>

	* sim/ucsim/pdk.src/inst.cc:
	  Fix addc m, a, patch by Nicolas Lesser,
	  fix subs m, a.
	* src/pdk/gen.c:
	  +=, -= optimization; fix wide == between stack and register operand.

2019-04-17 Philipp Klaus Krause <pkk AT spth.de>

	* sim/ucsim/pdk.src/pdk.cc:
	  Fix register dump, patch by Nicolas Lesser.

2019-04-17 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* src/pdk/gen.c:
	  Various fixes and improvements.
	* src/SDCCpeeph.c,
	  src/pdk/peeph.def:
	  Remove untargeted labels.
	* support/regression/tests/*.c:
	  Disable tests that won't work anytime soon on pkd15.
	* sim/ucsim/globals.cc:
	  More verbose pdk subtype output.

2019-04-16 Philipp Klaus Krause <pkk AT spth.de>

	* Merged pdk branch to trunk.

2019-04-15 Philipp Klaus Krause <pkk AT spth.de>

	* sdas/linksrc/lkrloc3.c:
	  pdk13 linker fix, patch By Nicolas Lesser.

2019-04-15 Philipp Klaus Krause <pkk AT spth.de>

	* Merged post-3.9 branch to trunk.

2019-04-15 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* src/pdk/gen.c:
	  Fix POINTER_SET with left operand on stack, right operand in pseudo-register.

2019-04-14 Philipp Klaus Krause <pkk AT spth.de>

	* support/regression/tests/bug-460010.c,
	  support/regression/tests/bug1928022.c,
	  support/regression/tests/gcc-torture-execute-991216-2.c,
	  support/regression/tests/gcc-torture-execute-va-arg-1.c,
	  support/regression/tests/gcc-torture-execute-va-arg-15.c:
	  Disable some tests for pdk15, on which they won't work anytime soon.
	* src/pdk/gen.c:
	  Improve handling of 16-bit return value,
	  fix stack parameters,
	  fix carry in comparison of wide stack values.
	* src/pdk/peeph.def:
	  Remove double exchange.

2019-04-13 Philipp Klaus Krause <pkk AT spth.de>

	* src/pdk/gen.c:
	  Fix a bug in 16-bit store from registers via rematerialized pointer with non-zero offset;
	  fix a bug in handling of on-stack operands to &,|;
	  fix a bug in handling of past-end offsets for stack operands.

2019-04-09 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* src/pdk/gen.c:
	  Fix handling of -0.0 in ifx.

2019-04-09 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* sdas/linksrc/lkrloc3.c:
	  Linker warning improvement, patch by Nicolas Lesser.
	* support/regression/tests/*.c:
	  Disable some tests for pdk15, on which they won't work anytime soon.
	* src/pdk/gen.c:
	  Fix wide left shift by large literal.
	* sim/ucsim/pdk.src/glob.cc,
	  sim/ucsim/pdk.src/glob.h,
	  sim/ucsim/pdk.src/inst.cc,
	  sim/ucsim/pdk.src/instcl.h,
	  sim/ucsim/pdk.src/pdk.cc:
	  pdk13 simulator, patch by Nicolas Lesser.

2019-04-08 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* support/regression/tests/*.c:
	  Disable some tests for pdk15, on which they won't work anytime soon.
	* sdas/linksrc/aslink.h,
	  sdas/linksrc/lkmain.c,
	  sdas/linksrc/lkrloc3.c:
	  Linker memory size warning fixes, patches by Nicolas Lesser.
	* src/pdk/ralloc.c:
	  Fix handling of shift result assigned to sfr.

2019-04-06 Philipp Klaus Krause <pkk AT spth.de>

	* sdas/aspdk13/pdk13pst.c,
	  sdas/aspdk13/pdk13mch.c:
	  Assembler fixes.

2019-04-05 Philipp Klaus Krause <pkk AT spth.de>

	* sdas/aspdk13/pdk13pst.c,
	  sdas/aspdk13/pdk13mch.c:
	  Assembler fixes.

2019-04-05 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* sdas/aspdk13/pdk13mch.c,
	  sdas/aspdk13/pdk13pst.c:
	  Assembler fix, patch by Leo.
	* sim/ucsim/pdk.src/pdk.cc:
	  Simulator fix, patch by Nicolas Lesser.

2019-04-05 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* .version,
	  doc/INSTALL.txt,
	  sdcc.spec:
	  Bump version in preparation of 3.9.0 release.

2019-04-05 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* support/regression/tests/*:
	  Disable some tests for pdk15, on which they won't work anytime soon.
	* sdas/aspdk15/pdk15mch.c
	  Assembler fix, patch by Nicolas Lesser.
	* sim/ucsim/pdk.src/inst.cc,
	  sim/ucsim/pdk.src/pdk.cc,
	  sim/ucsim/pdk.src/regspdk.h:
	  Simulator fix, patch by Nicolas Lesser.
	* sdas/aspdk13/pdk13mch.c:
	  Assembler fix, patch by Leo.

2019-04-05 Philipp Klaus Krause <krauseph AT informatik.uni-freiburg.de>

	* src/pdk/gen.c,
	  support/regression/tests/bug-2274.c:
	  Support dummy write for handling of uninitialized variables, better code for write of literal to bit-field.
	* sim/ucsim/pdk.src/inst.cc:
	  Fix dec m for pdk15.
	* support/regression/tests/bitfields.c,
	  support/regression/tests/bug-2181.c,
	  support/regression/tests/bug-2197.c,
	  support/regression/tests/bug-221100.c,

src/caltech10/gen.c
/*-------------------------------------------------------------------------
  gen.c - source file for code generation for the Caltech10
-------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "common.h"
#include "ralloc.h"
#include "gen.h"
#include "dbuf_string.h"

/** 1 if result is rematerializable */
static int resultRemat(iCode * ic) {
  if (SKIP_IC (ic) || ic->op == IFX)
    return 0;

  if (IC_RESULT (ic) && IS_ITEMP (IC_RESULT (ic))) {
    symbol *sym = OP_SYMBOL (IC_RESULT (ic));
    if (sym->remat && !POINTER_SET (ic))
      return 1;
  }
  return 0;
}


/*****************************************************************************/
/* asmop helpers                                                             */
/*****************************************************************************/

#define AOP(op) op->aop
#define AOP_TYPE(op) AOP(op)->type
#define AOP_SIZE(op) AOP(op)->size
#define AOP_OP(aop) aop->op

/** Create new asmop */
static asmop *newAsmop (short type) {
  asmop *aop = Safe_calloc (1, sizeof (asmop));
  aop->type = type;
  aop->op = NULL;
  return aop;
}


/** Get asmop of a true symbol */
static asmop *aopForSym (iCode * ic, symbol * sym, bool result) {
  asmop *aop;
  memmap *space;

  wassertl (ic != NULL, "Got a null iCode");
  wassertl (sym != NULL, "Got a null symbol");

//  printf("in aopForSym for symbol %s\n", sym->name);

  space = SPEC_OCLS (sym->etype);

  /* if already has one */
  if (sym->aop)
    return sym->aop;

  /* special case for a function */
  if (IS_FUNC (sym->type)) {
    sym->aop = aop = newAsmop(AOP_IMMD);
    aop->aopu.aop_immd.aop_immd1 = Safe_calloc (1, strlen (sym->rname) + 1);
    strcpy (aop->aopu.aop_immd.aop_immd1, sym->rname);
    aop->size = FARPTRSIZE;
    return aop;
  }

  /* if it is on the stack */
  if (sym->onStack) {
    sym->aop = aop = newAsmop(AOP_SOF);
    aop->aopu.aop_dir = sym->rname;
    aop->size = getSize (sym->type);
    aop->aopu.aop_stk = sym->stack;
    return aop;
  }

  /* if it is in direct space */
  if (IN_DIRSPACE(space)) {
    sym->aop = aop = newAsmop (AOP_DIR);
    aop->aopu.aop_dir = sym->rname;
    aop->size = getSize (sym->type);
    return aop;
  }

  /* default to far space */
  sym->aop = aop = newAsmop (AOP_EXT);
  aop->aopu.aop_dir = sym->rname;
  aop->size = getSize (sym->type);
  return aop;
}


/** Allocate asmop for operand */
static void aopOp (operand *op, iCode * ic, bool result) {
  asmop *aop = NULL;
  symbol *sym;

  if (!op)
    return;

//  printf("checking literal\n");
  /* if this a literal */
  if (IS_OP_LITERAL (op)) {
    op->aop = aop = newAsmop (AOP_LIT);
    aop->aopu.aop_lit = OP_VALUE (op);
    aop->size = getSize (operandType (op));
    aop->op = op;
    return;
  }

//  printf("checking pre-existing\n");
  /* if already has a asmop then continue */
  if (op->aop) {
    op->aop->op = op;
    return;
  }

//  printf("checking underlying sym\n");
  /* if the underlying symbol has a aop */
  if (IS_SYMOP (op) && OP_SYMBOL (op)->aop) {
    op->aop = aop = Safe_calloc (1, sizeof (*aop));
    memcpy (aop, OP_SYMBOL (op)->aop, sizeof (*aop));
    //op->aop = aop = OP_SYMBOL (op)->aop;
    aop->size = getSize (operandType (op));
    //printf ("reusing underlying symbol %s\n",OP_SYMBOL (op)->name);
    //printf (" with size = %d\n", aop->size);

    aop->op = op;
    return;
  }

//  printf("checking true sym\n");
  /* if this is a true symbol */
  if (IS_TRUE_SYMOP (op)) {
    op->aop = aop = aopForSym (ic, OP_SYMBOL (op), result);
    aop->op = op;
    //printf ("new symbol %s\n", OP_SYMBOL (op)->name);
    //printf (" with size = %d\n", aop->size);
    return;
  }

  /* this is a temporary : this has
     only five choices :
     a) register
     b) spillocation
     c) rematerialize
     d) conditional
     e) can be a return use only */

  if (!IS_SYMOP(op))
    piCode(ic, NULL);
  sym = OP_SYMBOL(op);

//  printf("checking conditional\n");
  /* if the type is a conditional */
  // if (sym->regType == REG_CND)
  //   {
  //     sym->aop = op->aop = aop = newAsmop (AOP_CRY);
  //     aop->size = 0;
  //     aop->op = op;
  //     return;
  //   }
  // Check removed, because registers not supported

//  printf("checking spilt\n");
  /* if it is spilt then two situations
     a) is rematerialize
     b) has a spill location */
  if (sym->isspilt || sym->nRegs == 0)
    {
//      printf("checking remat\n");
      /* rematerialize it NOW */
      wassertl(sym->remat, "Rematerializing not supported");

      wassertl(!sym->ruonly, "sym->ruonly not supported");

      /* else spill location  */
      if (sym->isspilt && sym->usl.spillLoc)
        {
          asmop *oldAsmOp = NULL;

          if (sym->usl.spillLoc->aop && sym->usl.spillLoc->aop->size != getSize (sym->type))
            {
              /* force a new aop if sizes differ */
              oldAsmOp = sym->usl.spillLoc->aop;
              sym->usl.spillLoc->aop = NULL;
              //printf ("forcing new aop\n");
            }
          sym->aop = op->aop = aop = aopForSym (ic, sym->usl.spillLoc, result);
          if (sym->usl.spillLoc->aop->size != getSize (sym->type))
            {
              /* Don't reuse the new aop, go with the last one */
              sym->usl.spillLoc->aop = oldAsmOp;
            }
          aop->size = getSize (sym->type);
          aop->op = op;
          //printf ("spill symbol %s\n", OP_SYMBOL (op)->name);
          //printf (" with size = %d\n", aop->size);
          return;
        }

      /* else must be a dummy iTemp */
      sym->aop = op->aop = aop = newAsmop (AOP_DUMMY);
      aop->size = getSize (sym->type);
      aop->op = op;
      return;
    }

//  printf("assuming register\n");
  /* must be in a register */
  wassert(sym->nRegs);
  wassertl(0, "Register operand not supported");
}


/** Free an operand's asmop, or aaop itself, whichever is non-null. */
static void freeAsmop (operand * op, asmop * aaop, iCode * ic, bool pop) {
  asmop *aop;
  if (!op)
    aop = aaop;
  else
    aop = op->aop;

  if (!aop)
    return;

  if (aop->freed)
    goto dealloc;

  aop->freed = 1;

  if (aop->stacked)
    {
      wassertl(0, "Stacked aop not supported");
      // int stackAdjust;
      // int loffset;
      // aop->stacked = 0;
      // stackAdjust = 0;
      // for (loffset = 0; loffset < aop->size; loffset++)
      //   if (aop->stk_aop[loffset])
      //     {
      //       transferAopAop (aop->stk_aop[loffset], 0, aop, loffset);
      //       stackAdjust++;
      //     }
      // pullNull (stackAdjust);
    }

dealloc:
  /* all other cases just dealloc */
  if (op)
    {
      op->aop = NULL;
      if (IS_SYMOP (op))
        {
          OP_SYMBOL (op)->aop = NULL;
          /* if the symbol has a spill */
          if (SPIL_LOC (op))
            SPIL_LOC (op)->aop = NULL;
        }
    }
}

/*****************************************************************************/
/* Generate code for specific operations                                     */
/*****************************************************************************/

/** Load from aop */
static void loadFromAop(asmop *aop) {
  wassert(aop->size == 1);
  if (aop->type == AOP_LIT) {
    emitcode("LDI", "%02x", byteOfVal(aop->aopu.aop_lit, 0));
    return;
  }
  if (aop->type == AOP_DIR) {
    emitcode("LDD", "$%s", aop->aopu.aop_dir);
    return;
  }
  // Support literal and direct only
  wassertl(0, "Unsupported aop type");
}

/** Store to aop */
static void storeToAop(asmop *aop) {
  wassert(aop->size == 1);
  if (aop->type == AOP_DIR) {
    emitcode("STD", "$%s", aop->aopu.aop_dir);
    return;
  }
  // Support storing to direct only
  wassertl(0, "Unsupported aop type");
}

/** Generate code that copies from one operand to another */
static void genCopy(operand *result, operand *source) {
  int size = AOP_SIZE(result);
  int srcsize = AOP_SIZE(source);
  wassert(size == 1); // No support for big types yet
  wassert(srcsize == 1);

  // Only support copying from these types.
  wassert(AOP_TYPE(source) == AOP_LIT || AOP_TYPE(source) == AOP_DIR);
  loadFromAop(AOP(source));
  // Only support copying to direct
  wassert(AOP_TYPE(result) == AOP_DIR);
  storeToAop(AOP(result));
}

/** Generate code for assign */
static void genAssign(iCode *ic) {
  operand *result = IC_RESULT(ic);
  operand *right = IC_RIGHT(ic);
  aopOp(right, ic, FALSE);
  aopOp(result, ic, TRUE);
  genCopy(result, right);
  freeAsmop(right, NULL, ic, TRUE);
  freeAsmop(result, NULL, ic, TRUE);
}


static void genLabel(iCode *ic) {
  // Special case, never generate
  if (IC_LABEL (ic) == entryLabel)
    return;
  emitLabel(IC_LABEL(ic));
}

static void genFunction(iCode *ic) {
}


static void genReturn(iCode *ic) {
}


static void genEndFunction(iCode *ic) {
}


/*****************************************************************************/
/* Overall code generation                                                   */
/*****************************************************************************/

/** Generate code for a single iCode */
void genCaltech10iCode(iCode *ic) {
  if (resultRemat(ic) || ic->generated) {
    wassertl(0, "Rematerializable icode");
  }

  initGenLineElement();
  genLine.lineElement.ic = ic;

  switch (ic->op) {
    case '=':
      if (POINTER_SET (ic))
        wassertl(0, "Pointer set not supported");
      genAssign(ic);
      break;
    case LABEL:
      genLabel(ic);
      break;
    case FUNCTION:
      genFunction(ic);
      break;
    case RETURN:
      genReturn(ic);
      break;
    case ENDFUNCTION:
      genEndFunction(ic);
      break;
    default:
      piCode(ic, NULL);
      wassertl(0, "unrecognized iCode");
  }
}

/** Generate code for list of iCodes */
void genCaltech10Code (iCode *lic)
{
  int cln = 0; // Current line #

  // Print the allocation information
  if (currFunc)
    printAllocInfo(currFunc, codeOutBuf);

  // Mark all ics as not generated
  for (iCode *ic = lic; ic; ic = ic->next)
    ic->generated = FALSE;

  for (iCode *ic = lic; ic; ic = ic->next) {
    initGenLineElement ();
    genLine.lineElement.ic = ic;
    if (ic->lineno && cln != ic->lineno) {
      if (!options.noCcodeInAsm) {
        emitcode("", ";%s:%d: %s", ic->filename, ic->lineno, printCLine (ic->filename, ic->lineno));
      }
      cln = ic->lineno;
    }
    if (options.iCodeInAsm) {
      const char *iLine = printILine (ic);
      emitcode("", "; ic:%d: %s", ic->seq, iLine);
      dbuf_free (iLine);
    }
    genCaltech10iCode(ic);
  }
  // Now do the actual printing
  printLine(genLine.lineHead, codeOutBuf);

  // Destroy the line list
  destroy_line_list();
}



src/port.h
/** @file port.h
    Definitions for what a port must provide.
    All ports are referenced in SDCCmain.c.
 */
#ifndef PORT_INCLUDE
#define PORT_INCLUDE

#include "SDCCicode.h"
#include "SDCCargs.h"
#include "SDCCpeeph.h"
#include "dbuf.h"

#define TARGET_ID_MCS51    1
#define TARGET_ID_GBZ80    2
#define TARGET_ID_Z80      3
#define TARGET_ID_AVR      4
#define TARGET_ID_DS390    5
#define TARGET_ID_PIC14    6
#define TARGET_ID_PIC16    7
#define TARGET_ID_DS400    10
#define TARGET_ID_HC08     11
#define TARGET_ID_Z180     12
#define TARGET_ID_R2K      13
#define TARGET_ID_R3KA     14
#define TARGET_ID_S08      15
#define TARGET_ID_STM8     16
#define TARGET_ID_TLCS90   17
#define TARGET_ID_EZ80_Z80 18
#define TARGET_ID_PDK13    19
#define TARGET_ID_PDK14    20
#define TARGET_ID_PDK15    21
#define TARGET_ID_PDK16    22
#define TARGET_ID_CALTECH10 23

/* Macro to test the target we are compiling for.
   Can only be used after SDCCmain has defined the port
 */
#define TARGET_IS_MCS51    (port->id == TARGET_ID_MCS51)
#define TARGET_IS_AVR      (port->id == TARGET_ID_AVR)
#define TARGET_IS_DS390    (port->id == TARGET_ID_DS390)
#define TARGET_IS_DS400    (port->id == TARGET_ID_DS400)
#define TARGET_IS_PIC14    (port->id == TARGET_ID_PIC14)
#define TARGET_IS_PIC16    (port->id == TARGET_ID_PIC16)
#define TARGET_IS_Z80      (port->id == TARGET_ID_Z80)
#define TARGET_IS_Z180     (port->id == TARGET_ID_Z180)
#define TARGET_IS_R2K      (port->id == TARGET_ID_R2K)
#define TARGET_IS_R3KA     (port->id == TARGET_ID_R3KA)
#define TARGET_IS_GBZ80    (port->id == TARGET_ID_GBZ80)
#define TARGET_IS_TLCS90   (port->id == TARGET_ID_TLCS90)
#define TARGET_IS_EZ80_Z80 (port->id == TARGET_ID_EZ80_Z80)
#define TARGET_IS_HC08     (port->id == TARGET_ID_HC08)
#define TARGET_IS_S08      (port->id == TARGET_ID_S08)
#define TARGET_IS_STM8     (port->id == TARGET_ID_STM8)
#define TARGET_IS_PDK13    (port->id == TARGET_ID_PDK13)
#define TARGET_IS_PDK14    (port->id == TARGET_ID_PDK14)
#define TARGET_IS_PDK15    (port->id == TARGET_ID_PDK15)
#define TARGET_IS_PDK16    (port->id == TARGET_ID_PDK16)
#define TARGET_IS_CALTECH10 (port->id == TARGET_ID_CALTECH10)

#define TARGET_MCS51_LIKE  (TARGET_IS_MCS51 || TARGET_IS_DS390 || TARGET_IS_DS400)
#define TARGET_Z80_LIKE    (TARGET_IS_Z80 || TARGET_IS_Z180 || TARGET_IS_GBZ80 || TARGET_IS_R2K || TARGET_IS_R3KA || TARGET_IS_TLCS90 || TARGET_IS_EZ80_Z80)
#define TARGET_IS_RABBIT   (TARGET_IS_R2K || TARGET_IS_R3KA)
#define TARGET_HC08_LIKE   (TARGET_IS_HC08 || TARGET_IS_S08)
#define TARGET_PIC_LIKE    (TARGET_IS_PIC14 || TARGET_IS_PIC16)
#define TARGET_PDK_LIKE    (TARGET_IS_PDK13 || TARGET_IS_PDK14 || TARGET_IS_PDK15 || TARGET_IS_PDK16)

/* is using sdas / sdld assembler / linker */
#define IS_SDASLD          (TARGET_Z80_LIKE || TARGET_MCS51_LIKE || TARGET_HC08_LIKE)

#define MAX_BUILTIN_ARGS        16
/* definition of builtin functions */
typedef struct builtins
{
  char *name;                   /* name of builtin function */
  char *rtype;                  /* return type as string : see typeFromStr */
  int nParms;                   /* number of parms : max 8 */
  char *parm_types[MAX_BUILTIN_ARGS];   /* each parm type as string : see typeFromStr */
} builtins;

struct ebbIndex;

/* pragma structure */
struct pragma_s
{
  const char *name;
  int id;
  char deprecated;
  int (*func) (int id, const char *name, const char *cp);
};

/* defined in SDCClex.lex */
int process_pragma_tbl (const struct pragma_s *pragma_tbl, const char *s);

/* Processor specific names */
typedef struct
{
  /** Unique id for this target */
  const int id;
  /** Target name used for -m */
  const char *const target;

  /** Target name string, used for --help */
  const char *const target_name;

  /** Specific processor for the given target family. specified by -p */
  char *processor;

  struct
  {
    /** Pointer to glue function */
    void (*do_glue) (void);
    /** TRUE if all types of glue functions should be inserted into
        the file that also defines main.
        We dont want this in cases like the z80 where the startup
        code is provided by a seperate module.
     */
    bool glue_up_main;
    /* OR of MODEL_* */
    int supported_models;
    int default_model;
    /** return the model string, used as library destination;
        port->target is used as model string if get_model is NULL */
    const char *(*get_model) (void);
  }
  general;

  /* assembler related information */
  struct
  {
    /** Command to run and arguments (eg as-z80) */
    const char **cmd;
    /** Alternate macro based form. */
    const char *mcmd;
    /** Arguments for debug mode. */
    const char *debug_opts;
    /** Arguments for normal assembly mode. */
    const char *plain_opts;
    /* print externs as global */
    int externGlobal;
    /* assembler file extension */
    const char *file_ext;
    /** If non-null will be used to execute the assembler. */
    void (*do_assemble) (set *);
  }
  assembler;

  /* linker related info */
  struct
  {
    /** Command to run (eg link-z80) */
    const char **cmd;
    /** Alternate macro based form. */
    const char *mcmd;
    /** If non-null will be used to execute the link. */
    void (*do_link) (void);
    /** Extension for object files (.rel, .obj, ...) */
    const char *rel_ext;
    /** 1 if port needs the .lnk file, 0 otherwise */
    const int needLinkerScript;
    const char *const *crt;
    const char *const *libs;
  }
  linker;

  /** Default peephole rules */
  struct
  {
    char *default_rules;
    int (*getSize) (lineNode * line);
    bitVect *(*getRegsRead) (lineNode * line);
    bitVect *(*getRegsWritten) (lineNode * line);
    bool (*deadMove) (const char *reg, lineNode * currPl, lineNode * head);
    bool (*notUsed) (const char *reg, lineNode * currPl, lineNode * head);
    bool (*canAssign) (const char *op1, const char *op2, const char *op3);
    bool (*notUsedFrom) (const char *reg, const char *label, lineNode *head);
    bool (*symmParmStack) (void);
  }
  peep;

  /** Basic type sizes */
  struct
  {
    int char_size;
    int short_size;
    int int_size;
    int long_size;
    int longlong_size;
    int near_ptr_size;          // __near
    int far_ptr_size;           // __far
    int ptr_size;               // generic
    int funcptr_size;
    int banked_funcptr_size;
    int bit_size;
    int float_size;
  }
  s;

  /** tags for far, near, xstack, code generic pointers */
  struct
  {
    int tag_far;
    int tag_near;
    int tag_xstack;
    int tag_code;
  }
  gp_tags;

  /** memory regions related stuff */
  struct
  {
    const char *const xstack_name;
    const char *const istack_name;
    /*
     * The following 2 items can't be const pointers
     * due to ugly implementation in gbz80 target;
     * this should be fixed in src/z80/main.c (borutr)
     */
    const char *code_name;
    const char *data_name;
    const char *const idata_name;
    const char *const pdata_name;
    const char *const xdata_name;
    const char *const bit_name;
    const char *const reg_name;
    const char *const static_name;
    const char *const overlay_name;
    const char *const post_static_name;
    const char *const home_name;
    const char *const xidata_name;      // initialized xdata
    const char *const xinit_name;       // a code copy of xidata
    const char *const const_name;       // const data (code or not)
    const char *const cabs_name;        // const absolute data (code or not)
    const char *const xabs_name;        // absolute xdata/pdata
    const char *const iabs_name;        // absolute idata/data
    const char *const initialized_name; // Initialized global (and static local) variables.
    const char *const initializer_name; // A code copy of initialized_name (to be copied for fast initialization).
    struct memmap *default_local_map;   // default location for auto vars
    struct memmap *default_globl_map;   // default location for globl vars
    int code_ro;                        // code space read-only 1=yes
    unsigned int maxextalign;           // maximum extended alignment supported, nonnegative power of 2 (C11 standard, section 6.2.8).
  }
  mem;

  struct
  {
    void (*genExtraAreaDeclaration) (FILE *, bool);
    void (*genExtraAreaLinkOptions) (FILE *);
  }
  extraAreas;

  /* stack related information */
  struct
  {
    /** -1 for grows down (z80), +1 for grows up (mcs51) */
    int direction;
    /** Extra overhead when calling between banks */
    int bank_overhead;
    /** Extra overhead when the function is an ISR */
    int isr_overhead;
    /** Standard overhead for a function call */
    int call_overhead;
    /** Re-enterant space */
    int reent_overhead;
    /** 'banked' call overhead.
        Mild overlap with bank_overhead */
    int banked_overhead;
    /** 0 if sp points to last item pushed, 1 if sp points to next location to use */
    int offset;
  }
  stack;

  struct
  {
    /** Size of the biggest shift the port can handle. -1 if port can handle shifts of arbitrary size. */
    signed int shift;

    /* Has support routines for int x int -> long multiplication and unsigned int x unsigned int -> unsigned long multiplication */
    bool has_mulint2long;
  }
  support;

  struct
  {
    void (*emitDebuggerSymbol) (const char *);
    struct
    {
      int (*regNum) (const struct reg_info *);
      bitVect *cfiSame;
      bitVect *cfiUndef;
      int addressSize;
      int regNumRet;
      int regNumSP;
      int regNumBP;
      int offsetSP;
    }
    dwarf;
  }
  debugger;

  struct
  {
    int maxCount;
    int sizeofElement;
    int sizeofMatchJump[3];
    int sizeofRangeCompare[3];
    int sizeofSubtract;
    int sizeofDispatch;
  }
  jumptableCost;

  /** Prefix to add to a C function (eg "_") */
  const char *fun_prefix;

  /** Called once the processor target has been selected.
      First chance to initalise and set any port specific variables.
      'port' is set before calling this.  May be NULL.
  */
  void (*init) (void);
  /** Parses one option + its arguments */
  bool (*parseOption) (int *pargc, char **argv, int *i);
  /** Optional list of automatically parsed options.  Should be
      implemented to at least show the help text correctly. */
  OPTION *poptions;
  /** Initialise port spectific paths */
  void (*initPaths) (void);
  /** Called after all the options have been parsed. */
  void (*finaliseOptions) (void);
   /** Called after the port has been selected but before any
       options are parsed. */
  void (*setDefaultOptions) (void);
  /** Does the dirty work. */
  void (*assignRegisters) (struct ebbIndex *);

  /** Returns the register name of a symbol.
      Used so that 'reg_info' can be an incomplete type. */
  const char *(*getRegName) (const struct reg_info *reg);

  int (*getRegByName) (const char *name);

  /** Try to keep track of register contents. */
  bool (*rtrackUpdate)(const char* line);

  /* list of keywords that are used by this
     target (used by lexer) */
  char **keywords;

  /* Write any port specific assembler output. */
  void (*genAssemblerPreamble) (FILE * of);
  /* invoked at end assembler file */
  void (*genAssemblerEnd) (FILE * of);

  /* Write the port specific IVT. If genIVT is NULL or if
   * it returns zero, default (8051) IVT generation code
   * will be used.
   */
  int (*genIVT) (struct dbuf_s * oBuf, symbol ** intTable, int intCount);

  void (*genXINIT) (FILE * of);

  /* Write port specific startup code */
  void (*genInitStartup) (FILE * of);

  /* parameter passing in register related functions */
  void (*reset_regparms) (struct sym_link *);        /* reset the register count */
  int (*reg_parm) (struct sym_link *, bool reentrant);  /* will return 1 if can be passed in register */

  /** Process the pragma string 'sz'.  Returns 0 if recognised and
      processed, 1 otherwise.  May be NULL.
   */
  int (*process_pragma) (const char *sz);

  /** Mangles a support function name to reflect the calling model.
   */
  const char *(*getMangledFunctionName) (const char *szOrginial);

  /** Returns true if the port can multiply the two types nativly
      without using support functions.
   */
  bool (*hasNativeMulFor) (iCode *ic, sym_link *left, sym_link *right);

  /** Returns true if the port has implemented certain bit
      manipulation iCodes (RRC, RLC, SWAP, GETHBIT, GETABIT, GETBYTE, GETWORD)
   */
  bool (*hasExtBitOp) (int op, int size);

  /** Returns the relative expense of accessing a particular output
      storage class. Larger values indicate higher expense.
   */
  int (*oclsExpense) (struct memmap * oclass);

  /** If TRUE, then tprintf and !dw will be used for some initalisers
   */
  bool use_dw_for_init;

  /** TRUE for targets with little endian byte ordering, FALSE for
      targets with big endian byte ordering.
   */
  bool little_endian;

  /* condition transformations */
  bool lt_nge;                  /* transform (a < b)  to !(a >= b)  */
  bool gt_nle;                  /* transform (a > b)  to !(a <= b)  */
  bool le_ngt;                  /* transform (a <= b) to !(a > b)   */
  bool ge_nlt;                  /* transform (a >= b) to !(a < b)   */
  bool ne_neq;                  /* transform a != b --> ! (a == b)  */
  bool eq_nne;                  /* transform a == b --> ! (a != b)  */

  bool arrayInitializerSuppported;
  bool (*cseOk) (iCode * ic, iCode * pdic);
  builtins *builtintable;       /* table of builtin functions */
  int unqualified_pointer;      /* unqualified pointers type is  */
  int reset_labelKey;           /* reset Label no 1 at the start of a function */
  int globals_allowed;          /* global & static locals not allowed ?  0 ONLY TININative */

  int num_regs;                 /* Number of registers handled in the tree-decomposition-based register allocator in SDCCralloc.hpp */

#define PORT_MAGIC 0xAC32
  /** Used at runtime to detect if this structure has been completely filled in. */
  int magic;
}
PORT;

extern PORT *port;

#if !OPT_DISABLE_MCS51
extern PORT mcs51_port;
#endif
#if !OPT_DISABLE_Z80
extern PORT z80_port;
#endif
#if !OPT_DISABLE_Z180
extern PORT z180_port;
#endif
#if !OPT_DISABLE_R2K
extern PORT r2k_port;  /* Rabbit 2000/3000 */
#endif
#if !OPT_DISABLE_R3KA
extern PORT r3ka_port; /* Rabbit 3000A */
#endif
#if !OPT_DISABLE_GBZ80
extern PORT gbz80_port;
#endif
#if !OPT_DISABLE_TLCS90
extern PORT tlcs90_port;
#endif
#if !OPT_DISABLE_EZ80_Z80
extern PORT ez80_z80_port;
#endif
#if !OPT_DISABLE_AVR
extern PORT avr_port;
#endif
#if !OPT_DISABLE_DS390
extern PORT ds390_port;
#endif
#if !OPT_DISABLE_PIC14
extern PORT pic_port;
#endif
#if !OPT_DISABLE_PIC16
extern PORT pic16_port;
#endif
#if !OPT_DISABLE_TININative
extern PORT tininative_port;
#endif
#if !OPT_DISABLE_DS400
extern PORT ds400_port;
#endif
#if !OPT_DISABLE_HC08
extern PORT hc08_port;
#endif
#if !OPT_DISABLE_S08
extern PORT s08_port;
#endif
#if !OPT_DISABLE_STM8
extern PORT stm8_port;
#endif
#if !OPT_DISABLE_PDK13
extern PORT pdk13_port;
#endif
#if !OPT_DISABLE_PDK14
extern PORT pdk14_port;
#endif
#if !OPT_DISABLE_PDK15
extern PORT pdk15_port;
#endif
#if !OPT_DISABLE_CALTECH10
extern PORT caltech10_port;
#endif

#endif /* PORT_INCLUDE */
